File: balanced_group_system/balanced_group_system.py
import random
from . import group_system

"""
BalancedGroupSystem.py

This module contains a subclass of GroupSystem that creates balanced groups based on the familiarity matrix.
It creates groups in such a way that members meet as many new members as possible and minimize repeat meetings.

Author: Joshua Si
"""

class BalancedGroupSystem(group_system.GroupSystem):
  def __init__(self, members: list[str] = []):
    super().__init__(members)
    self.familiarity_matrix: dict[frozenset[tuple[str, str]], int] = {
        frozenset((self.members[i], self.members[j])) : 0
        for i in range(len(self.members))
        for j in range(i+1, len(self.members))
    }

  def __repr__(self) -> str:
    return f"BalancedGroupSystem({self.members})"
  
  def add_member(self, member: str) -> None:
    super().add_member(member)
    for i in range(len(self.members)-1):
      self.familiarity_matrix[frozenset((self.members[i], member))] = 0
  
  def remove_member(self, member: str) -> None:
    try:
      super().remove_member(member)
      for m in self.members:
        self.familiarity_matrix.pop(frozenset((m, member)))
    except ValueError:
      raise ValueError("Member '{}' was not in members".format(member))
    
  def create_groups(self, groups: list[set[str]]) -> list[set[str]]:
    super().create_groups(groups)
    for group in groups:
      self._update_familiarity(group)
    return groups

  def print_familiarity(self) -> None:
    for i in self.members:
      print(i, end=' ')
    print()
    for i in self.members:
      print(i, end=' ')
      for j in self.members:
        if i == j:
          print('-', end=' ')
        else:
          print(self.familiarity_matrix[frozenset((i, j))], end=' ')
      print()

  def evaluate_group(self, group: set[str]) -> int:
    score = 0
    pairs = ((i, j) for i in group for j in group if i != j)
    score = sum(self.familiarity_matrix[frozenset(pair)] for pair in pairs)
    return score
  
  '''
  Update familiarity matrix with given group. Used internally when creating groups.
  '''
  def _update_familiarity(self, group: set[str]) -> None:
    pairs = ((i, j) for i in group for j in group if i != j)
    for pair in pairs:
      self.familiarity_matrix[frozenset(pair)] += 1

  '''
  Calculate a possible set of balanced groups by trying to minimize score when choosing which group to add each member to.
  Dry run equivalent to create_balanced_groups that does not update group history or familiarity matrix.
  '''
  def _calculate_balanced_groups(self, group_count: int, members: list[str] = [], verbose: bool = False) -> list[set[str]]:
    random.shuffle(members)
    groups = [set() for _ in range(group_count)]
    scores = {(): 0}
    # Memoized version of evaluate_group for evaluating score when adding one member
    def evaluate_member(group: set[str], member: str) -> int:
      new_group: set[str] = group | {member}
      if frozenset(new_group) in scores:
        return scores[frozenset(new_group)]
      score = scores[frozenset(group)] if frozenset(group) in scores else 0
      for i in group:
        score += self.familiarity_matrix[frozenset((i, member))]
      scores[frozenset(new_group)] = score + 2
      return score + 2

    # Balance and distribute members
    for member in members:
      min_group_index = min(range(group_count), key=lambda i: evaluate_member(groups[i], member))
      groups[min_group_index].add(member)
      if verbose:
        print(groups)
    return groups


  def create_balanced_groups(self, group_count: int, verbose: bool = False) -> list[set[str]]:
    groups = self._calculate_balanced_groups(group_count, self.members.copy(), verbose)
    return self.create_groups(groups)

File: balanced_group_system/balanced_group_system.cpp
#include "balanced_group_system.hpp"
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <random>

BalancedGroupSystem::BalancedGroupSystem(const std::vector<std::string>& members) : GroupSystem(members), familiarity_matrix(members.size(), std::vector<int>(members.size(), 0)) {
}

void BalancedGroupSystem::add_member(const std::string& member) {
    GroupSystem::add_member(member);
    for (auto& row : familiarity_matrix) {
        row.push_back(0);
    }
    familiarity_matrix.push_back(std::vector<int>(familiarity_matrix.size(), 0));
}

void BalancedGroupSystem::remove_member(const std::string& member) {
    auto it = std::find(members.begin(), members.end(), member);
    if (it == members.end()) {
        throw std::runtime_error("Member not found");
    }
    int index = std::distance(members.begin(),it);
    GroupSystem::remove_member(member);
    for (auto& row : familiarity_matrix) {
        row.erase(row.begin() + index);
    }
    familiarity_matrix.erase(familiarity_matrix.begin()+index);
}

void BalancedGroupSystem::create_groups(const std::vector<std::vector<std::string>>& group_list) {
    GroupSystem::create_groups(group_list);
    for (const auto& group : group_list) {
        update_familiarity(group);
    }
}

void BalancedGroupSystem::print_familiarity() {
    std::cout << "  ";
    for(std::string member : members) {
        std::cout << member << ", ";
    }
    std::cout << std::endl;
    for (const auto& row : familiarity_matrix) {
        for (auto const& element : row) {
            std::cout << element << " ";
        }
        std::cout << std::endl;
    }
}

int BalancedGroupSystem::evaluate_group(const std::vector<std::string>& group) {
    int score = 0;
    for (size_t i = 0; i < group.size(); ++i) {
        for (size_t j = i + 1; j < group.size(); ++j) {
            score += familiarity_matrix[std::distance(members.begin(), std::find(members.begin(), members.end(), group[i]))][std::distance(members.begin(), std::find(members.begin(), members.end(), group[j]))];
        }
    }
    return score;
}

void BalancedGroupSystem::update_familiarity(const std::vector<std::string>& group) {
    for (size_t i = 0; i < group.size(); ++i) {
        for (size_t j = i + 1; j < group.size(); ++j) {
            familiarity_matrix[std::distance(members.begin(), std::find(members.begin(), members.end(), group[i]))][std::distance(members.begin(), std::find(members.begin(), members.end(), group[j]))] += 1;
            familiarity_matrix[std::distance(members.begin(), std::find(members.begin(), members.end(), group[j]))][std::distance(members.begin(), std::find(members.begin(), members.end(), group[i]))] += 1;
        }
    }
}

std::vector<std::vector<std::string>> BalancedGroupSystem::calculate_balanced_groups(int group_count, std::vector<std::string> new_members) {
    std::random_shuffle(new_members.begin(), new_members.end());
    std::vector<std::vector<std::string>> groups(group_count, std::vector<std::string>());
    std::vector<int> scores;
    std::map<std::vector<std::string>, int> memoized_scores;
    
    auto evaluate_member = [&](std::vector<std::string> group, std::string member) {
        std::vector<std::string> group_copy = group;
        std::sort(group_copy.begin(), group_copy.end());
        std::vector<std::string> group_and_member;
        group_and_member.reserve(group_copy.size() + 1);
        group_and_member = group_copy;
        group_and_member.push_back(member);
        std::sort(group_and_member.begin(), group_and_member.end());

        if (memoized_scores.find(group_and_member) != memoized_scores.end()){
            return memoized_scores[group_and_member];
        }
        int score = 0;
        for (std::string i : group) {
             score += familiarity_matrix[std::distance(members.begin(), std::find(members.begin(), members.end(), i))][std::distance(members.begin(), std::find(members.begin(), members.end(), member))];
        }
        memoized_scores[group_and_member] = score + 1;
        return score + 1;
    };

    for (std::string member : new_members) {
        int min_group_index = 0;
        for (int i = 0; i < group_count; i++) {
            if (evaluate_member(groups[min_group_index], member) < evaluate_member(groups[i], member)){
                min_group_index = i;
            }
        }
        groups[min_group_index].push_back(member);
    }
    return groups;
}

std::vector<std::vector<std::string>> BalancedGroupSystem::create_balanced_groups(int group_count) {
    std::vector<std::string> copy_members = members;
    return calculate_balanced_groups(group_count, copy_members);
}
